### ISSUE TITLE
RasterLite2 - How to build sample

### ISSUE THREAD

**Issue by mj10777** [geopaparazzi/geopaparazzi#276]
If you have a nice GeoTiff that you would like to done
* send it to me and I will return a step by step sample how this should be done

This then then be used a base to get it done for your tools.

**Comment by moovida** [comment]
Amazing, thanks. Would it be possible to do that with a set of tiffs? I.e. with an image mosaic (maybe done with gdal? If I remember correctly, RasterLite2 doesn't save things as tiles, but in some smart multiresolution mode? That would make it possible to have even large large datasets at the disk cost of the original data or a bit more. Against tiles, that get into terabytes if the region is too large and the zoomlevels too high.

I am heading out now. I will come back on this, thanks!

**Comment by mj10777** [comment]
You can import a Directoy of files with one command
* something like `RL2_GetMapImageFromRasteFromDir`

**Comment by mj10777** [comment]
On further thought, this would be interesting to try again
* with the very first version, there were sometimes gaps between the files

It would be good idea to check if this still happens.
So if you have, lets say, 9 images (3x3 rows), 1 image being surrounded by the other 8
* I could check if these gaps still occur
   * and if yes, report it to Sando

**Comment by moovida** [comment]
Ok, I am uploading two datasets and sending them to you in private.

* ortofoto.zip: this one contains 9 tiles (3x3) in epsg:32632. The tiffs are tiffs with worldfile tfw and are 3-band images.
* chm_geotiff.zip: this on contains 6 tiles (3x2) in epsg:32632. These are single band geotiffs. They represent the difference between the DSM and the DTM. You will see many trees on it. Since they are single band carrying a physical information, an sld style file is attached in order to be styled (see attached image for what it should look like). I wonder if it is possible to have in rasterlite both information, i.e. be able to get the height of a tree, but also visualize the image of this. If that is possible in rasterlite2, this is my new favourite format also for desktop apps :-)

![mark_picture](https://cloud.githubusercontent.com/assets/390250/8300726/1f31818c-1988-11e5-816e-598b9772d9e1.PNG)

**Comment by mj10777** [comment]
2015-06-23 9:13 GMT+02:00 Andrea Antonello <notifications@github.com>:

> Ok, I am uploading two datasets and sending them to you in private.
>
>    - ortofoto.zip: this one contains 9 tiles (3x3) in epsg:32632. The
>    tiffs are tiffs with worldfile tfw and are 3-band images.
>    - chm_geotiff.zip: this on contains 6 tiles (3x2) in epsg:32632. These
>    are single band geotiffs. They represent the difference between the DSM and
>    the DTM. You will see many trees on it. Since they are single band carrying
>    a physical information, an sld style file is attached in order to be styled
>    (see attached image for what it should look like). I wonder if it is
>    possible to have in rasterlite both information, i.e. be able to get the
>    height of a tree, but also visualize the image of this. If that is possible
>    in rasterlite2, this is my new favourite format also for desktop apps :-)
>
> If not, I am sure Sandro will make an attempt to make it so.

Is this similar?
https://www.gaia-gis.it/fossil/librasterlite2/wiki?name=trento-ctr

**Comment by moovida** [comment]
> Is this similar?
> https://www.gaia-gis.it/fossil/librasterlite2/wiki?name=trento-ctr

No, that one is a monocrome, which contains either black or white, i.e. 0 or 1. 
The CHM example has floating point values and we should be able to represent it in two ways:
* the physical: elevation value in one point. That would make it possible to even query a map on geopap (chuckle, chuckle...) or make some small analyses. :-)
* the image: image tiles to show as background.

**Comment by mj10777** [comment]
2015-06-23 9:23 GMT+02:00 Andrea Antonello <notifications@github.com>:

> Is this similar?
> https://www.gaia-gis.it/fossil/librasterlite2/wiki?name=trento-ctr
>
> No, that one is a monocrome, which contains either black or white, i.e. 0
> or 1.
> The CHM example has floating point values and we should be able to
> represent it in two ways:
>
>    - the physical: elevation value in one point. That would make it
>    possible to even query a map on geopap (chuckle, chuckle...) or make some
>    small analyses. :-)
>    - the image: image tiles to show as background.
>
>
ans were have they been uploaded to?

**Comment by moovida** [comment]
It is at 77%, you will get the email when it is done.

**Comment by moovida** [comment]
Sandro showed me this (a bit old, so he says) doc about single band raster styling: https://www.gaia-gis.it/fossil/librasterlite2/wiki?name=ETOPO1

**Comment by mj10777** [comment]
OK, had a quick look.
* your sld files look a bit different than those of his sample

We shall see what happens, when I get that far
* had to deal with another matter this morning, so am a bit behind

**Comment by moovida** [comment]
> OK, had a quick look.
> 
> your sld files look a bit different than those of his sample
> We shall see what happens, when I get that far

ok, it is a standard, so it should be fine either way. Not sure what spatialite uses for sld evaluation though.

> had to deal with another matter this morning, so am a bit behind

No problem at all. I am on other work, so i won't be on that for a little while.

**Comment by mj10777** [comment]
There is something wrong (they do not show up in qgis) with:
* uni_bz_52, 58 and 59

![qgis_ortofoto_error](https://cloud.githubusercontent.com/assets/4899405/8323753/f289346a-1a48-11e5-95e1-f51868475f71.png)

**Comment by mj10777** [comment]
I'm afraid that page seems to be bit more than old
* the logic has changed to much
   * the method to register the Style to the Raster is clear (SE_RegisterRasterStyledLayer)
      *  for coverage_name, from style_name

But how/where the style is stored
* is not clear reading the .c source code

I will need more information
* either prepared documentation (possible not yet completed)
* or sql samples on how it is done
   * maybe somewhere in the test files

Sandro should give us a hint on to continue on this.

**Comment by mj10777** [comment]
See note above about images not showing up
* still open

**Comment by moovida** [comment]
Hi @mj10777 , I forgot and got burried in other work. Also I set Sandro the same link and he imported all the rasters. Did you try to donwload the files again to check if they got somehow corrupted.

I am hard in creating the user manual for the upcoming FOSS4G Europe this week, sorry.

**Comment by mj10777** [comment]
Importing is ok, but how do they look - it should be a square.
I am also busy updating the maps due to new knowledge of certain positions.
But will be starting wiki-pages for each process of rasterlite2 creation
* when I get to styles and vector layers
   * I will be asking Sando for one or two samples
      * reading the code is not helpful, since it is combination of many things

If you contact Sando, ask if he has some simple samples
* I will then work it out

**Comment by mj10777** [comment]
Forget the last message
* all of these style commands are documented in spatialite and not in rasterlite2

**Comment by mj10777** [comment]
It seems that the .sld files you sent are of version 1.0.0 and not 1.1.0 that is being supported by spatialite and rasterlite2
* these files **cannot** be read

https://groups.google.com/forum/?utm_source=digest&utm_medium=email#!topic/spatialite-users/POly5Hxitbg

So one question would be if the software that created the .sld of chm_geotiff.zip can be set to use 1.1.0?

**Comment by mj10777** [comment]
For the `chm_geotiff.zip` portion I have some results:

First the bad news:
* the styles **must** use the 1.1.0 syntax
* the `QGIS` sln 1.1.0 styles **cannot** be used
* `geopaparazzi` must be adapted for this to show
   * no idea yet how, but that will be the next step

Second the good news:
* `spatialite_gui` will offer a simple way to create the needed styles
   * it took me about 5 minutes to create the needed `rl2se.DIFF1.3_2012_geo_1m.xml`
      * reading from the original .sln, which is similar

```xml
<?xml version="1.0" encoding="UTF-8"?>
<RasterSymbolizer version="1.1.0" xsi:schemaLocation="http://www.opengis.net/se http://schemas.opengis.net/se/1.1.0/Symbolizer.xsd" xmlns="http://www.opengis.net/se" xmlns:ogc="http://www.opengis.net/ogc" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<Name>rl2se.DIFF1.3_2012_geo_1m</Name>
	<Description>
		<Title>Andrea 'Geopaparazzi' DEM Sample</Title>
		<Abstract>for use of 'converting' a sld 1.0.0 to 1.1.1. Using 'RasterSymbolizer: Color Map - Categorize'</Abstract>
	</Description>
	<Opacity>1.00</Opacity>
	<ColorMap>
		<Categorize fallbackValue="#ffffff">
			<LookupValue>Rasterdata</LookupValue>
			<Value>#000000</Value>
			<Threshold>1.000000</Threshold>
			<Value>#ffff00</Value>
			<Threshold>20.799999</Threshold>
			<Value>#00ff00</Value>
			<Threshold>40.599998</Threshold>
			<Value>#00ffff</Value>
			<Threshold>60.399998</Threshold>
			<Value>#0000ff</Value>
			<Threshold>80.199997</Threshold>
			<Value>#ff00ff</Value>
			<Threshold>100.000000</Threshold>
			<Value>#ff0000</Value>
		</Categorize>
	</ColorMap>
</RasterSymbolizer>
```

![rl2se diff1 3_2012_geo_1m](https://cloud.githubusercontent.com/assets/4899405/8717773/bb3df556-2b9e-11e5-8bfe-600ea0526ead.png)

The final result shown in LibreWMS looks like this 
* screenshot Left: the WMS output, Tight: the image you sent me

![20150716 wmslite_original](https://cloud.githubusercontent.com/assets/4899405/8717819/0fcfacf4-2b9f-11e5-8d31-32df3ff4779a.png)

**Comment by QDPeng** [comment]
I want to know how to load the geoTiff image file into mapview by geopaparzzi,and it can zoom the tiff image file in mapview? thank you!
